feat: Add .NET getting started guide (#1)
testcontainers/tc-guide-getting-started-with-testcontainers-for-dotnet: guide/getting-started-with-testcontainers-for-dotnet/index.adoc

---
title: "Getting started with Testcontainers for .NET"

> date: 2023-02-10T09:39:58+05:30

Reviewer: DateTime needs to be updated. Or is it intended to show this guide in the beginning?

Author: Dunno, anys suggestions?

Reviewer: We can use the actual date of writing the guide.

Author: As discussed I've updated it to yesterday.

draft: false
description: This guide will help you to get started with Testcontainers for .NET by demonstrating how you can use PostgreSQL for testing.
repo: https://github.com/testcontainers/tc-guide-getting-started-with-testcontainers-for-dotnet
languages:

>   - C#

Reviewer: Should the language be ".NET"?

Author: No, .NET is the SDK or runtime. C# is the language, there is also VB.NET or  F#.

Let us look at how we can use Testcontainers to test a .NET application using a
Postgres database.

== Create a solution file with a source and test project

Create a .NET source and test project from your favorite IDE. Once the projects
are created, add the following dependencies to the source and test project.

.Source project
[source,xml,indent=0]
----
include::{codebase}/src/GettingStarted/GettingStarted.csproj[lines="6"]
----

.Test project
[source,xml,indent=0]
----
include::{codebase}/tests/GettingStarted.Tests/GettingStarted.Tests.csproj[lines="6..8"]
----

We have added *Npgsql*, a Postgres ADO.NET Data Provider for talking to the
Postgres database, and *xUnit.net* for testing our service.

== Implement business logic

We are going to create a `CustomerService` class to manage customer details.

First, let us create a `Customer` type as follows:

[source%nowrap,csharp]
----
include::{codebase}/src/GettingStarted/Customer.cs[]
----

Create a `DbConnectionProvider` class to hold ADO.NET connection parameters
(connection string) and create a method to get a database connection, as follows:

[source%nowrap,csharp]
----

> include::{codebase}/src/GettingStarted/DbConnectionProvider.cs[]

Reviewer: It should be "DBConnectionProvider.cs" (uppercase B).

Author: I favor `Db` to avoid [inconsistent name](https://www.jetbrains.com/help/rider/InconsistentNaming.html) warnings (probably idiomatic to the language and does not matter for the guide) in the IDE (UpperCamelCase / PascalCase). Our [modules](https://github.com/testcontainers/testcontainers-dotnet/tree/develop/src) follow the same style.

Reviewer: The asciidoctor will try to fetch the content from that file and it is case sensitive, so it is failing to include the code snippet. To fix that you can rename the file to "DbConnectionProvider.cs".

Author: Good catch 🦅 Sorry, I misunderstood that you meant the actual file. I've renamed the file that it works for case sensitive file systems too.

----

Create a `CustomerService` class and add the following code:

[source%nowrap,csharp]
----
include::{codebase}/src/GettingStarted/CustomerService.cs[]
----

Let us understand what is going on in the `CustomerService` class.

> * We are calling `_dbConnectionProvider.GetConnection()` to get a database
>   connection using ADO.NET.
> * We have the `CreateCustomersTable()` method that creates the `customers`
>   table if it does not already exist.
> * We have the `GetCustomers()` method that fetches all rows from the
>   `customers` table, populates data into `Customer` objects, and returns a list
>   of `Customer` objects.
> * We have the `Create(Customer)` method that inserts a new customer record into
>   the database.

Reviewer 2: ```suggestion
* `_dbConnectionProvider.GetConnection()` gets a database
  connection using ADO.NET.
* `CreateCustomersTable()` method creates the `customers`
  table if it does not already exist.
* `GetCustomers()` method fetches all rows from the
  `customers` table, populates data into `Customer` objects, and returns a list
  of `Customer` objects.
* `Create(Customer)` method inserts a new customer record into
  the database.
```

Now let us see how we can test the `CustomerService` logic using Testcontainers.

== Add Testcontainers dependencies

Before writing Testcontainers-based tests, let's add Testcontainers
dependencies to the test project as follows:

.Test project
[source,xml,indent=0]
----
include::{codebase}/tests/GettingStarted.Tests/GettingStarted.Tests.csproj[lines="9"]
----

As we are using a Postgres database for our application, we added the
Testcontainers Postgres

> https://www.nuget.org/packages/Testcontainers.PostgreSql[module] as a test

Reviewer: Nit: Add the link for "Postgres module" instead of just "module" as follows:

Testcontainers `https://www.nuget.org/packages/Testcontainers.PostgreSql[Postgres module]` as a test dependency.

Author: I've changed the link text.

> Testcontainers Postgres
> https://www.nuget.org/packages/Testcontainers.PostgreSql[module] as a test

Author: ```suggestion
Testcontainers https://www.nuget.org/packages/Testcontainers.PostgreSql[Postgres module] as a test
```

dependency.

== Write test using Testcontainers

Create a `CustomerServiceTest` class in the test project with the following
code:

[source%nowrap,csharp]
----
include::{codebase}/tests/GettingStarted.Tests/CustomerServiceTest.cs[]
----

Let us understand the code in our `CustomerServiceTest` class.

> * We declare `PostgreSqlContainer` by passing the Docker image name
>   `postgres:15-alpine` to the Postgres builder.
> * The Postgres container is started using xUnit.net's `IAsyncLifetime`
>   interface, which executes `InitializeAsync` immediately after the test class
>   has been created.
> * We have a `ShouldReturnTwoCustomers()` test where we initialize
>   `CustomerService`, insert two customer records into the database, fetch all the
>   existing customers, and assert the number of customers.
> * Finally, we dispose of the Postgres container in the `DisposeAsync()` member,
>   which gets executed after the test method is executed.

Reviewer 2: ```suggestion
* Declare `PostgreSqlContainer` by passing the Docker image name
  `postgres:15-alpine` to the Postgres builder.
* The Postgres container is started using xUnit.net's `IAsyncLifetime`
  interface, which executes `InitializeAsync` immediately after the test class
  has been created.
* `ShouldReturnTwoCustomers()` test initializes
  `CustomerService`, insert two customer records into the database, fetch all the
  existing customers, and assert the number of customers.
* Finally, the Postgres container is disposed in the `DisposeAsync()` member,
  which gets executed after the test method is executed.
```

> If you run the customer service test, you can see in the IDE's debug output

Reviewer 2: ```suggestion
By running the customer service test, you can see in the IDE's debug output
```
